Add after_discard method to ActiveJob

cardy31 opened the pull request:
`after_discard` lets job authors define a block which will be run when a job is about to be discarded.

This has utility for both job authors and for gems/modules included on jobs, which can use this method to run a block when a job fails.

`after_discard` respects the existing retry behaviour of jobs, but will run even if a retried exception is handled in a block passed to `retry_on` or `discard_on`.

<!--
Thanks for contributing to Rails!

Please do not make *Draft* pull requests, as they still send
notifications to everyone watching the Rails repo.

Create a pull request when it is ready for review and feedback
from the Rails team :).

If your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

About this template

The following template aims to help contributors write a good description for their pull requests.
We'd like you to provide a description of the changes in your pull request (i.e. bugs fixed or features added), motivation behind the changes, and complete the checklist below before opening a pull request.

Feel free to discard it if you need to (e.g. when you just fix a typo). -->

### Motivation / Background

<!--
Describe why this Pull Request needs to be merged. What bug have you fixed? What feature have you added? Why is it important?
If you are fixing a specific issue, include "Fixes #ISSUE" (replace with the issue number, remove the quotes) and the issue will be linked to this PR.
-->

`after_discard` has been used in Shopify's monolith for many years. Developers take advantage of it to define cleanup and failure handling code, and features have been built that rely on it for tracking job failures.

While other parts of ActiveJob's current exception handling system get close to this behaviour, I believe that there is value in having a clearly defined way to run a block of code when a job has failed.

### Detail

This Pull Request changes no existing behaviour, but adds `after_discard`.

`after_discard` is called in any situation where the job is about to be discarded due to a failure. Those cases are:

- When an exception is being raised by the job without `retry_on` or `discard_on` used
- When an exception is handled by `discard_on` (with or without a block)
- When an exception is handled by `retry_on` (with or without a block), but only after the job has run out of retries for that exception
- If multiple exceptions are raised when running the `after_discard` blocks, then only the last exception will be raised. This behaviour ensures that all blocks will be run despite exceptions.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

rafaelfranca commented on activejob/test/jobs/final_failure_retry_job.rb:
```diff
@@ -0,0 +1,34 @@
+require_relative "../support/job_buffer"
+require "active_support/core_ext/integer/inflections"
+
+class UnhandledError < StandardError; end
+class DefaultsError < StandardError; end
+class CustomCatchError < StandardError; end
+class DiscardableError < StandardError; end
+class CustomDiscardableError < StandardError; end
+class FinalFailureError < StandardError; end
+class ChildFinalFailureError < FinalFailureError; end
```
Let's avoid defining top level constants in this test to not clash with others. Let's move to inside the job.

rafaelfranca commented on activejob/test/jobs/final_failure_retry_job.rb:
```diff
@@ -0,0 +1,34 @@
+
+class FinalFailureRetryJob < ActiveJob::Base
+  retry_on DefaultsError
+  retry_on(CustomCatchError) { |job, error| JobBuffer.add("Dealt with a job that failed to retry in a custom way after #{job.arguments.second} attempts. Message: #{error.message}") }
```
Are we going to assert this message?

cardy31 replied:
I'll add the assertion

cardy31 commented on activejob/lib/active_job/exceptions.rb:
```diff
@@ -127,6 +147,12 @@ def retry_job(options = {})
       end
     end
 
+    def run_final_failure_procs(exception)
```
I'll make it private

bensheldon commented:
As a backend maintainer, I love this 😍 Two things:

- I would like to see an option _not_ to further raise the exception to the backend. It's really, really tough ([example](https://github.com/bensheldon/good_job/issues/159#issuecomment-848545857) from @morgoth) to set up retries with error-reporting services and not end up with duplicate exception reports when the backend is _also_ reporting exceptions (and sometimes even handling retries itself). It would be really helpful to be able to use this as an (optional) comprehensive blanket to handle exceptions within a job. 
- I don't like that the name introduce a new concept of "final" and "failure", which I don't think are clear given that different backends might subsequently handle it in different ways. Pedantically it looks like this is invoked on discards, "retry stopped", and unrescued errors. I want to suggest a more generic `on_error` or `after_error` handler. Naming is hard 😓

simi commented:
Wouldn't this be better as `after_discard` method? If I understand it well, the terminology is following. Started job could result into retry or end as succeeded or discarded.

cardy31 commented:
@bensheldon The current approach of the PR is to simply raise the error in every case where it would already be raised by ActiveJob.

A different approach would be to follow the pattern used by `retry_on` where the exception is passed to the block. If the job author would like to raise the error beyond that block, they have to re-raise within the block. This would mean that when using `on_final_failure` in a job, it would change existing raising behaviour of `retry_on` and `discard_on`

Alternatively, as suggested we could have an option on the `on_final_failure` method to raise. Something like `on_final_failure(raise_after_block = false)`. This is less in line with the current exception handling patterns, but communicates what will happen to developers quite clearly.

I don’t have particularly strong opinions about which way to go, so I’d love to know what you and the community think would be best.

@simi I agree that `after_discard` is more consistent with the current ActiveJob naming. I'll update the PR to use that. 👍

bensheldon commented:
One weak quibble with the name `after_discard`: The implementation would imply that `retry_stopped` == `discard`. I think [that's currently ambiguous](https://github.com/bensheldon/good_job/issues/844). I really like your proposal because it feels like a blanket, side-effect free way to prevent exceptions being raised to the backend (that's my interest here), but I think it widens the definition of "discard" significantly and that makes me uncomfortable (though not blocking, just more like "discard means this now. so say we all") 

> Alternatively, as suggested we could have an option on the `after_discard` method to raise. Something like `after_discard(raise_after_block = false)`. This is less in line with the current exception handling patterns, but communicates what will happen to developers quite clearly.

This would be my preference, with a kwarg like `after_discard(raise_to_backend: false)` which I think would be consistent with Active Job terminology. 

... that led me to ask the question "what happens if a `throw :abort` happens?", which then made me realize that this isn't implemented as a Active Support Callback, and is slightly different than `rescue_from`, though similar to `retry_on`'s block.

I think that anything that is named `before_`/`after_`/`around_` should be implemented similar to Active Support Callbacks, executing in the context of the instances rather than having the instance passed as an argument to the block. So maybe that puts in favor of `on_discard` though I'm still unsure again of whether `retry_stopped` or an implicit exception is equivalent to `discard` 🤷🏻 Unfortunately I don't have strong opinions either, so feel free to override me ❤️

cardy31 committed:
This method lets job authors define a block which will be run when a job is about to be discarded.

This has utility for gems/modules included on jobs, which can tie into this behaviour and run something when a job fails.
after_discard respects the existing retry behaviour, but will run even if a retried exception is handled in a block.
